tests/abi/tools/revng-ensure-rft-equivalence/Main.cpp

```diff
@@ -74,60 +103,159 @@ int main(int Argc, char *Argv[]) {
   if (EC)
     ExitOnError(llvm::createStringError(EC, EC.message()));
 
-  // Gather `RawFunctionType`s present in both the models based on their IDs.
+  // Introduce a function pair container to allow grouping them based on their
+  // `CustomName`.
   struct FunctionPair {
-    model::RawFunctionType *Left;
-    model::RawFunctionType *Right;
+    model::RawFunctionType *Left = nullptr;
+    model::RawFunctionType *Right = nullptr;
   };
-  std::map<model::Type::IDType, FunctionPair> Functions;
-  for (model::UpcastableType &LeftType : LeftModel->Model->Types())
-    if (auto *Left = llvm::dyn_cast<model::RawFunctionType>(LeftType.get()))
-      Functions[Left->ID()].Left = Left;
-  for (model::UpcastableType &RightType : RightModel->Model->Types())
-    if (auto *Right = llvm::dyn_cast<model::RawFunctionType>(RightType.get()))
-      Functions[Right->ID()].Right = Right;
-
-  // Ensure their stack arguments have the same ID.
-  //
-  // If two `RawFunctionType`s have the same ID but their stack argument structs
-  // do not - replace the ID of the "Right" struct to be that of the "Left" one.
-  std::map<model::TypePath, model::TypePath> Replacements;
-  for (auto [_, Pair] : Functions) {
+  struct TransparentComparator {
+    using is_transparent = std::true_type;
+
+    bool operator()(llvm::StringRef LHS, llvm::StringRef RHS) const {
+      return LHS < RHS;
+    }
+  };
+  std::map<model::Identifier, FunctionPair, TransparentComparator> Functions;
+  std::unordered_set<std::size_t> FunctionIDLookup;
+
+  // Make sure the default prototype is valid.
+  revng_assert(LeftModel->Model->DefaultPrototype().isValid());
+  const auto &DefaultPrototype = *LeftModel->Model->DefaultPrototype().get();
+
+  // Gather all the `RawFunctionType`s prototypes present in the first model.
+  for (model::Function &LF : LeftModel->Model->Functions()) {
+    if (LF.Prototype().get() == nullptr)
+      continue; // Skip functions without prototypes.
+
+    revng_assert(LF.name() != "",
+                 "This test uses names to differenciate functions, as such "
+                 "having unnamed functions in the model would break it, "
+                 "hence it's not allowed.");
+
+    auto *Left = llvm::dyn_cast<model::RawFunctionType>(LF.Prototype().get());
+    revng_assert(Left != nullptr);
+    FunctionIDLookup.emplace(Left->ID());
+
+    if (Left->ID() == DefaultPrototype.ID())
+      continue; // Skip the default prototype.
+
+    auto [Iterator, Success] = Functions.try_emplace(LF.name());
+    if (!Success) {
+      std::string Error = "Two functions share the name in the left model: "
+                          + LF.name().str().str();
+      revng_abort(Error.c_str());
+    }
```

Comment:
Drop, `verify` already ensures this does not happen.
Same below.

Reply:
Dropped.